extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java

```diff
@@ -276,13 +318,124 @@ public KafkaIOConfig getIOConfig()
+                waitForPublishes.countDown();
+                break;
+              }
+
+              log.info("Publishing segments for sequence [%s]", sequenceMetadata);
+
+              final SegmentsAndMetadata result = driver.publish(
+                  sequenceMetadata.getPublisher(toolbox, ioConfig.isUseTransaction()),
+                  sequenceMetadata.getCommitterSupplier(topic, lastPersistedOffsets).get(),
+                  ImmutableList.of(sequenceMetadata.getSequenceName())
+              ).get();
+
+              if (result == null) {
+                throw new ISE(
+                    "Transaction failure publishing segments for sequence [%s]",
+                    sequenceMetadata
+                );
+              } else {
+                log.info(
+                    "Published segments[%s] with metadata[%s].",
+                    Joiner.on(", ").join(
+                        result.getSegments().stream().map(DataSegment::getIdentifier).collect(Collectors.toList())
+                    ),
+                    result.getCommitMetadata()
+                );
+              }
+
+              sequences.remove(sequenceMetadata);
+              publishingSequences.remove(sequenceMetadata.getSequenceName());
+              try {
+                persistState(toolbox);
+              }
+              catch (IOException e) {
+                log.error(e, "Unable to persist state, dying");
+                Throwables.propagate(e);
+              }
+
+              final ListenableFuture<SegmentsAndMetadata> handOffFuture = driver.registerHandoff(result);
+              handOffWaitList.add(handOffFuture);
+            }
+            catch (Throwable t) {
+              if ((t instanceof InterruptedException || (t instanceof RejectedExecutionException
+                                                         && t.getCause() instanceof InterruptedException))) {
+                log.warn("Stopping publish thread as we are interrupted, probably we are shutting down");
+              } else {
+                log.makeAlert(t, "Error in publish thread, dying").emit();
+                throwableAtomicReference.set(t);
+              }
+              Futures.allAsList(handOffWaitList).cancel(true);
+              waitForPublishes.countDown();
+              break;
+            }
+          }
+        }
+    );
+  }
+
   @Override
   public TaskStatus run(final TaskToolbox toolbox) throws Exception
   {
     log.info("Starting up!");
+
     startTime = DateTimes.nowUtc();
-    mapper = toolbox.getObjectMapper();
     status = Status.STARTING;
+    this.toolbox = toolbox;
+
+    if (getContext() != null && getContext().get("checkpoints") != null) {
+      log.info("Got checkpoints [%s]", (String) getContext().get("checkpoints"));
+      final TreeMap<Integer, Map<Integer, Long>> checkpoints = toolbox.getObjectMapper().readValue(
+          (String) getContext().get("checkpoints"), new TypeReference<TreeMap<Integer, Map<Integer, Long>>>()
```

Comment:
is it required for checkpoints map to be represented as string in context ? can't we have non-string structures stored in task context ?

Reply:
context can hold anything but as far as I remember there was serialization issue that's why I chose String. Will look at it again.